src/fromager/read.py

```diff
@@ -0,0 +1,22 @@
+import io
+import pathlib
+import typing
+from contextlib import contextmanager
+
+from .request_session import session
+
+
+@contextmanager
+def open_file_or_url(
+    path_or_url: str | pathlib.Path,
+) -> typing.Generator[io.TextIOWrapper, typing.Any, None]:
+    location = str(path_or_url)
+    if location.startswith(("https://", "http://", "file://")):
+        response = session.get(location)
```

Comment:
session.get() and open() will produce different types of exceptions, so every caller using this function would have to handle both exception types. That leaks details of the implementation outside of the function.

We can keep it like this in this MR, but it would be better to trap the exceptions and produce a common one that is advertised as the API for this function.

Reply:
Ah that makes sense

Reply:
Actually won't it help during debugging if the exceptions were different?

Reply:
If you do `raise IOError('local message') from err` then the caller only has to catch `IOError`, and the traceback shows all of the details of the original exception.